docs/docsite/rst/dev_guide/debugging.rst

```diff
@@ -0,0 +1,157 @@
+.. _debugging:
+
+=================
+Debugging Modules
+=================
+
+Debugging (local)
+=================
+
+To break into a module running on localhost and step through with the debugger:
+
+- Set a breakpoint in the module: ``import pdb; pdb.set_trace()``
+- Run the module on the local machine: ``$ python -m pdb ./my_new_test_module.py ./args.json``
+
+Debugging (remote)
+==================
+
+To debug a module running on a remote target (i.e. not localhost):
+
+- On your controller machine (running Ansible) set `ANSIBLE_KEEP_REMOTE_FILES=1` (this tells Ansible to retain the modules it sends to the remote machine instead of removing them)
+- Run your playbook targetting the remote machine and specify ``-vvvv`` (the verbose output will show you many things, including the remote location that Ansible uses for the modules)
+- Take note of the remote path Ansible used on the remote host
+- SSH into the remote target after the completion of the playbook
+- Navigate to the directory (most likely it is going to be your ansible remote user defined or implied from the playbook: ``~/.ansible/tmp/ansible-tmp-...``)
+- Here you should see the module that you executed from your Ansible controller, but this is the zipped file that Ansible sent to the remote host. You can run this by specifying ``python my_test_module.py`` (not necessary)
+- To debug, though, we will want to extract this zip out to the original module format: ``python my_test_module.py explode`` (Ansible will expand the module into ``./debug-dir``)
+- Navigate to ``./debug-dir`` (notice that unzipping has caused the generation of ``ansible_module_my_test_module.py``)
+- Modify or set a breakpoint in the unzipped module
+- Ensure that the unzipped module is executable: ``$ chmod 755 ansible_module_my_test_module.py``
+- Run the unzipped module directly passing the args file: ``$ ./ansible_module_my_test_module.py args`` (args is the file that contains the params that were originally passed. Good for repro and debugging)
+
+
+.. _debugging_ansiblemodule_based_modules:
+
+Debugging AnsibleModule-based modules
+`````````````````````````````````````
+
+.. tip::
+
+    If you're using the :file:`hacking/test-module` script then most of this
```

Comment:
Is `test-module` still working and maintained? I would check with @abadger because I'm not sure.